src/seq_mv/csr_matrix.c

```diff
@@ -33,79 +28,275 @@ hypre_CSRMatrixCreate( HYPRE_Int num_rows,
-   hypre_CSRMatrixNumNonzeros(matrix) = num_nonzeros;
+   hypre_CSRMatrixData(matrix)           = NULL;
+   hypre_CSRMatrixI(matrix)              = NULL;
+   hypre_CSRMatrixJ(matrix)              = NULL;
+   hypre_CSRMatrixBigJ(matrix)           = NULL;
+   hypre_CSRMatrixRownnz(matrix)         = NULL;
+   hypre_CSRMatrixNumRows(matrix)        = num_rows;
+   hypre_CSRMatrixNumRownnz(matrix)      = num_rows;
+   hypre_CSRMatrixNumCols(matrix)        = num_cols;
+   hypre_CSRMatrixNumNonzeros(matrix)    = num_nonzeros;
+   hypre_CSRMatrixMemoryLocation(matrix) = hypre_HandleMemoryLocation(hypre_handle());
 
    /* set defaults */
-   hypre_CSRMatrixOwnsData(matrix)  = 1;
-   if (num_nonzeros)
-   {
-      hypre_CSRMatrixNumRownnz(matrix) = num_rows;
-   }
-   else
-   {
-      hypre_CSRMatrixNumRownnz(matrix) = 0;
-   }
+   hypre_CSRMatrixOwnsData(matrix)       = 1;
+
+#if defined(HYPRE_USING_CUSPARSE)
+   hypre_CSRMatrixSortedJ(matrix)        = NULL;
+   hypre_CSRMatrixSortedData(matrix)     = NULL;
+   hypre_CSRMatrixCsrsvData(matrix)      = NULL;
+#endif
 
    return matrix;
 }
+
 /*--------------------------------------------------------------------------
  * hypre_CSRMatrixDestroy
  *--------------------------------------------------------------------------*/
 
 HYPRE_Int
 hypre_CSRMatrixDestroy( hypre_CSRMatrix *matrix )
 {
+   HYPRE_Int ierr = 0;
+
    if (matrix)
    {
-      hypre_TFree(hypre_CSRMatrixI(matrix));
-      hypre_CSRMatrixI(matrix) = NULL;
-      if (hypre_CSRMatrixRownnz(matrix))
-      {
-         hypre_TFree(hypre_CSRMatrixRownnz(matrix));
-      }
-      if (hypre_CSRMatrixOwnsData(matrix))
+      HYPRE_MemoryLocation memory_location = hypre_CSRMatrixMemoryLocation(matrix);
+
+      hypre_TFree(hypre_CSRMatrixI(matrix),      memory_location);
+      hypre_TFree(hypre_CSRMatrixRownnz(matrix), HYPRE_MEMORY_HOST);
+
+      if ( hypre_CSRMatrixOwnsData(matrix) )
       {
-         hypre_TFree(hypre_CSRMatrixData(matrix));
-         hypre_TFree(hypre_CSRMatrixJ(matrix));
-         hypre_CSRMatrixData(matrix) = NULL;
-         hypre_CSRMatrixJ(matrix)    = NULL;
+         hypre_TFree(hypre_CSRMatrixData(matrix), memory_location);
+         hypre_TFree(hypre_CSRMatrixJ(matrix),    memory_location);
+         /* RL: TODO There might be cases BigJ cannot be freed FIXME
+          * Not so clear how to do it */
+         hypre_TFree(hypre_CSRMatrixBigJ(matrix), memory_location);
       }
-      hypre_TFree(matrix);
-      matrix = NULL;
+
+#if defined(HYPRE_USING_CUSPARSE)
+      hypre_TFree(hypre_CSRMatrixSortedData(matrix), memory_location);
+      hypre_TFree(hypre_CSRMatrixSortedJ(matrix), memory_location);
+      hypre_CsrsvDataDestroy(hypre_CSRMatrixCsrsvData(matrix));
+#endif
+
+#if defined(HYPRE_USING_CUSPARSE) || defined(HYPRE_USING_ROCSPARSE)
+      hypre_GpuMatDataDestroy(hypre_CSRMatrixGPUMatData(matrix));
+#endif
+
+      hypre_TFree(matrix, HYPRE_MEMORY_HOST);
    }
 
-   return hypre_error_flag;
+   return ierr;
```

Comment:
Why doe we have 'ierr' here, when we used to have 'hypre_error_flag'? Why is ierr seet to 0?

Reply:
@ulrikeyang, I agree it should be `hypre_error_flag` as in the `recmat` branch. This change comes from master and there are other occurences like this in other functions. Should we change `return ierr` to `return hypre_error_flag` in `master` first and then update `recmat-merge` with the new `master`?

Reply:
I say yes, change it in master first.  However, I don't think we need to wait for the change on master to finalize this PR.  We will continue to update `recmat` by regularly pulling in changes from master anyway.  My 2 cents.